### PR TITLE
feat(daily-notes): extend moment→Luxon token map, reject all unsupported tokens

### PR DESCRIPTION
**Pull request by aliasunder**
## Summary

- Refactors the `momentToLuxonFormat` converter from sequential `reduce`+`replaceAll` to a single-pass regex tokenizer — tries tokens longest-first at each position, cleanly handling overlapping families (`DDDD`/`DDD`/`DD`/`D`) that the sequential approach could not
- Adds 4 new supported token mappings: `D`→`d`, `M`→`M`, `DDD`→`o`, `DDDD`→`ooo`
- Detects and rejects all Moment tokens that have no Luxon equivalent — ordinals (`Do`, `Mo`, `DDDo`, `wo`), `dd` (2-letter weekday), `d` (weekday number), `e`, `k`/`kk`, and the `L`-family localized formats (`L`–`LLLL`, `LT`, `LTS`). These would produce filenames that can never match what Obsidian writes:
  - **Env-var path:** warns at startup in `config.ts` (server keeps running; operator visibility)
  - **Both paths:** throws at `getDailyNotePath` with a clear error visible to MCP clients
- `findUnsupportedTokens` is standalone — uses its own boundary-guarded regex patterns, independent of the converter's token table

## Test plan

- [x] 8 new `momentToLuxonFormat` scenarios (new tokens, D-family/M-family ordering, adjacency)
- [x] 12 `findUnsupportedTokens` tests (ordinals, dd, d, e, k/kk, L-family, literals, ddd/dddd false-positive guard)
- [x] 2 config tests (Do and dd warn at startup)
- [x] 2 `getDailyNotePath` tests (Do and dd throw on path resolution)
- [x] 2 integration tests (server starts with unsupported format + tool call returns error)
- [x] 2 mutations verified (D↔Do ordering swap; DDD removal)
- [x] Full suite: 2790 tests pass
- [x] Manual Luxon probe: all new tokens render correctly

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_013uy4gPHmKzC9qBkTSeytBR

### CODE REVIEW

File: src/vault-mcp/obsidian-markdown/__tests__/moment-format.test.ts [hunk 1]
```diff
@@ -63,9 +63,78 @@ describe("momentToLuxonFormat", () => {
...
+    expect(hasOrdinalDayToken("YYYY-MM-DD")).toBe(false)
+  })
+
+  it("returns false when Do is inside a [literal] escape", () => {
+    expect(hasOrdinalDayToken("[Do] DD")).toBe(false)
+  })
+
+  it("returns true when Do is outside a literal even if a literal exists", () => {
+    expect(hasOrdinalDayToken("[Note] Do")).toBe(true)
+  })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/obsidian-markdown/__tests__/moment-format.test.ts]: Pre-existing behavior — `MOMENT_ESCAPE_RE` and its odd/even split pattern existed before this PR and are unchanged. Unterminated brackets in Moment format strings are extremely rare (Obsidian's UI enforces matched brackets). The existing tests cover the realistic boundaries; pathological literal edge cases are a separate concern from the token-mapping extension.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/obsidian-markdown/moment-format.ts [hunk 2]
```diff
@@ -1,43 +1,68 @@
...
   ["mm", "mm"],
   ["ss", "ss"],
+  // 1-char tokens
+  ["D", "d"], // day of month, no padding (1–31)
+  ["M", "M"], // month number, no padding (1–12) — identity, pinned
   ["A", "a"],
 ]
 
 /** Matches Moment.js [literal] escape groups — e.g. [Daily Note]. */
 const MOMENT_ESCAPE_RE = /\[([^\]]*)\]/g
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/obsidian-markdown/moment-format.ts:31]: Pre-existing behavior — `MOMENT_ESCAPE_RE` is unchanged by this PR. The regex has handled unbalanced brackets this way since its introduction, and Obsidian's date format UI produces matched brackets. A regex hardening pass for malformed formats is a separate concern from the token-mapping extension.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/__tests__/config.test.ts [hunk 3]
```diff
@@ -254,6 +254,11 @@ describe("loadConfig", () => {
         'env-var: "DAILY_NOTES_FORMAT" renders to an empty filename',
       )
     })
+
+    it("accepts a Do format without throwing (lossy but structurally safe)", () => {
+      const config = loadConfig({ DAILY_NOTES_FORMAT: "MMMM Do, YYYY" })
+      expect(config.dailyNotesFormat).toBe("MMMM Do, YYYY")
+    })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/__tests__/config.test.ts]: Already addressed — commit `5e3b8a2` (Phase 3 test audit) added both a positive test (`warns when format contains Do`) with a `logger.warn` spy asserting the exact message, and a negative test (`does not warn for a standard format without Do`). See `config.test.ts:264-278`.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/obsidian-markdown/moment-format.ts [hunk 4]
```diff
@@ -1,43 +1,68 @@
...
   ["DD", "dd"],
+  ["Do", "d"], // ordinal day — lossy: suffix ("st","nd","th") dropped
   ["HH", "HH"],
   ["hh", "hh"],
   ["mm", "mm"],
   ["ss", "ss"],
+  // 1-char tokens
+  ["D", "d"], // day of month, no padding (1–31)
+  ["M", "M"], // month number, no padding (1–12) — identity, pinned
   ["A", "a"],
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/obsidian-markdown/moment-format.ts:26]: Out of scope for this PR — the task note scoped this to `D`, `Do`, `DDD`, `DDDD`, and `M` (the tokens identified in the PR #429 review). `dd` is documented as unmapped in the JSDoc (`dd` — Moment's 2-letter weekday abbreviation). Other Moment tokens (`L`, `LL`, `w`, `WW`, `X`, `x`, `Z`) are not realistic daily-note filename formats. The README correctly says "most" tokens map, not "all".

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/obsidian-markdown/moment-format.ts [hunk 5]
```diff
@@ -1,43 +1,68 @@
...
+ *  single-pass regex — each position matches the longest token first. */
 const convertMomentTokens = (formatSpan: string): string =>
-  MOMENT_TO_LUXON.reduce(
-    (convertedSpan, [momentToken, luxonToken]) =>
-      convertedSpan.replaceAll(momentToken, luxonToken),
-    formatSpan,
+  formatSpan.replace(
+    MOMENT_TOKEN_RE,
+    (match) => MOMENT_TOKEN_MAP.get(match) ?? match,
   )
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/obsidian-markdown/moment-format.ts:46]: Acknowledged — by design. The single-pass regex reads the original format string, not its own output, so token-letter overlap in Luxon output is structurally impossible. The comment in the module docstring ("single-pass regex tokenizer") documents this guarantee.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/daily-notes.ts [hunk 6]
```diff
@@ -62,6 +65,11 @@ const readDailyNotesFileConfig = async (
           ? parsedConfig.format
           : FALLBACK_CONFIG.format,
     }
+    if (hasOrdinalDayToken(fileConfig.format)) {
+      logger.warn(
+        "daily-notes.json format contains Do (ordinal day) — the server will use the day number without suffix, so filenames will differ from Obsidian's",
+      )
+    }
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/daily-notes.ts]: Already addressed — commit `5e3b8a2` (Phase 3 test audit) added a test in `daily-notes.test.ts:131-148` that writes a `daily-notes.json` with a `Do` format, calls `readDailyNotesConfig`, and asserts the exact warning message via `logger.warn` spy.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/daily-notes.ts [hunk 7]
```diff
@@ -62,6 +65,11 @@ const readDailyNotesFileConfig = async (
           ? parsedConfig.format
           : FALLBACK_CONFIG.format,
     }
+    if (hasOrdinalDayToken(fileConfig.format)) {
+      logger.warn(
+        "daily-notes.json format contains Do (ordinal day) — the server will use the day number without suffix, so filenames will differ from Obsidian's",
+      )
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/daily-notes.ts]: Valid — fixed in `7e6a55f`. Moved the `hasOrdinalDayToken` warning from `readDailyNotesFileConfig` to `readDailyNotesConfig`, gated on `!envSettings?.format` so it only fires when the file's format is the effective one. Added a regression test asserting no warning when the env format overrides a `Do` file format.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/__tests__/config.test.ts [hunk 8]
```diff
@@ -254,6 +255,28 @@ describe("loadConfig", () => {
...
+
+    it("warns when format contains Do (ordinal day)", () => {
+      const warnSpy = vi.spyOn(logger, "warn").mockImplementation(() => {})
+      loadConfig({ DAILY_NOTES_FORMAT: "MMMM Do, YYYY" })
+      expect(warnSpy).toHaveBeenCalledTimes(1)
+      expect(warnSpy).toHaveBeenCalledWith(
+        "DAILY_NOTES_FORMAT contains Do (ordinal day) — Luxon has no ordinal-suffix token; the server will use the day number without suffix, so filenames will differ from Obsidian's",
+      )
+      warnSpy.mockRestore()
+    })
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/__tests__/config.test.ts]: Fixed in `ecd9e2b` — switched both config.test.ts and daily-notes.test.ts warn spies to use `onTestFinished(() => warnSpy.mockRestore())` so assertion failures cannot leak mocks to subsequent tests.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/daily-notes.ts [hunk 9]
```diff
@@ -88,6 +91,13 @@ export const readDailyNotesConfig = async (
...
 
   const fileConfig = await readDailyNotesFileConfig(vaultPath)
+  // Warn about Do only when the file format is the effective format —
+  // an env override makes the file's Do irrelevant.
+  if (!envSettings?.format && hasOrdinalDayToken(fileConfig.format)) {
+    logger.warn(
+      "daily-notes.json format contains Do (ordinal day) — the server will use the day number without suffix, so filenames will differ from Obsidian's",
+    )
+  }
   return {
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/daily-notes.ts]: Fixed in `ecd9e2b` — added a module-level `ordinalDayWarningEmitted` flag so the daily-notes.json Do warning fires at most once, regardless of how many times `readDailyNotesConfig` is called per session.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: README.md [hunk 10]
```diff
@@ -361,7 +361,9 @@ See [`templates/memory/`](./templates/memory/) for memory file examples and the
...
 
-When the file isn't available — or you use the Periodic Notes plugin, whose settings it doesn't reflect — set `DAILY_NOTES_FOLDER` (any vault-relative path: `Journal`, `Planner/Daily`) and `DAILY_NOTES_FORMAT` (same tokens as Obsidian's date format setting: `YYYY-MM-DD-dddd`, `YYYY/MM/DD`, …). You can set one or both — a set value always wins over the config file. Without either source, the server falls back to `Daily Notes` and `YYYY-MM-DD`.
+When the file isn't available — or you use the Periodic Notes plugin, whose settings it doesn't reflect — set `DAILY_NOTES_FOLDER` (any vault-relative path: `Journal`, `Planner/Daily`) and `DAILY_NOTES_FORMAT` (same tokens as Obsidian's date format setting: `YYYY-MM-DD-dddd`, `YYYY/MM/DD`, `MMM D, YYYY`, …). You can set one or both — a set value always wins over the config file. Without either source, the server falls back to `Daily Notes` and `YYYY-MM-DD`.
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply README.md:364]: False positive — the README has one paragraph at line 364, not two. The edit replaced the original line in place (adding `MMM D, YYYY` to the example list). The diff shows a removed and an added line because the content changed, but the result is a single paragraph.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/daily-notes.ts [hunk 11]
```diff
@@ -88,6 +94,18 @@ export const readDailyNotesConfig = async (
...
 
   const fileConfig = await readDailyNotesFileConfig(vaultPath)
+  // Warn about Do only when the file format is the effective format —
+  // an env override makes the file's Do irrelevant.
+  if (
+    !ordinalDayWarningEmitted &&
+    !envSettings?.format &&
+    hasOrdinalDayToken(fileConfig.format)
+  ) {
+    logger.warn(
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/daily-notes.ts]: False positive — the `ordinalDayWarningEmitted` flag and the `readDailyNotesConfig` warning were removed in `0369c6d`. The warning now lives only in `config.ts` (startup, env-var path); `daily-notes.ts` has no warning and no flag. This finding references code from an intermediate commit that no longer exists at HEAD.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
**Re: umm-actually PR-level finding — "Verify ARCHITECTURE.md daily-notes token claim"**

False positive. ARCHITECTURE.md line 196 describes the resolution chain (`DAILY_NOTES_FOLDER`/`DAILY_NOTES_FORMAT` env → `.obsidian/daily-notes.json` → fallback) and does not claim Moment.js token parity. The "same tokens as Obsidian's date format setting" phrasing is in the README, where it's accurate — Obsidian uses Moment tokens and we support most of them, with `Do` and `dd` explicitly documented as unsupported.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

File: src/vault-mcp/vault-operations/daily-notes.ts [hunk 12]
```diff
@@ -109,6 +112,14 @@ export const getDailyNotePath = async (params: {
 }): Promise<string> => {
   const { vaultPath, date, envSettings } = params
   const config = await readDailyNotesConfig(vaultPath, envSettings)
+
+  const unsupportedTokens = findUnsupportedTokens(config.format)
+  if (unsupportedTokens.length > 0) {
+    throw new Error(
+      `daily note format contains unsupported token(s): ${unsupportedTokens.join(", ")} — Luxon cannot reproduce what Obsidian writes, so the server would never find the note; change the format in Obsidian or set DAILY_NOTES_FORMAT to a supported format`,
+    )
+  }
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/vault-operations/daily-notes.ts:121]: Valid — the PR description was stale after the warn→throw refactor. Updated the summary to accurately describe the two paths: env-var warns at startup (config.ts), both paths throw at `getDailyNotePath` (tool-call surface visible to MCP clients).

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*

**Comment by aliasunder** [comment]
@coderabbit review

File: src/vault-mcp/config.ts [hunk 13]
```diff
@@ -36,10 +40,10 @@ const splitCommaSeparatedValues = (raw: string): string[] =>
...
 
 /** Validates a DAILY_NOTES_FORMAT value by probe-rendering a fixed date.
- *  Structural checks only — moment tokens without a Luxon mapping pass
- *  through and may render differently than in Obsidian, so only
- *  structurally unsafe results (traversal, separators, empty) are
- *  rejected. Returns the raw moment string unchanged. */
+ *  Structural checks only — structurally unsafe results (traversal,
+ *  separators, empty) are rejected. Warns when the format contains
+ *  unsupported tokens (Do, dd) whose Luxon mappings produce filenames
+ *  differing from Obsidian's. Returns the raw moment string unchanged. */
```
[the remark this answers is not in the document]
Comment (aliasunder) [reply src/vault-mcp/config.ts]: Valid — the tokens you listed (`d`, `e`, `k`/`kk`, `L`-family) were already added to `findUnsupportedTokens` in `8e2f6a5` (before this review ran). `dabcbbd` adds the remaining four: `Q` (quarter), `w` (locale week), `Z`/`ZZ` (timezone offsets). Tokens that pass through correctly (`E`, `X`, `x`, `W`, `S`/`SSS`) are not in the unsupported set since Luxon interprets them identically to Moment.

---
*🔍 ship-check · pr-monitor · Claude Opus 4.6 (1M context)*
